cadence/contracts/DapperStorageRent.cdc

```diff
@@ -0,0 +1,247 @@
+      let vaultRef = self.account.borrow<&FlowToken.Vault>(from: /storage/flowTokenVault)
+      if vaultRef == nil {
+        emit RefilledFailed(address: address, reason: "Couldn't borrow the Accounts FlowToken.Vault")
+        return
+      }
+
+      let privateForwardingSenderRef = self.account.borrow<&PrivateReceiverForwarder.Sender>(from: PrivateReceiverForwarder.SenderStoragePath)
+      if privateForwardingSenderRef == nil {
+        emit RefilledFailed(address: address, reason: "Couldn't borrow the Accounts PrivateReceiverForwarder")
+        return
+      }
+
+      // Check to make sure the payment vault has sufficient funds
+      if let vaultBalanceRef = self.account.getCapability(/public/flowTokenBalance).borrow<&FlowToken.Vault{FungibleToken.Balance}>() {
+        if vaultBalanceRef.balance <= REFUEL_AMOUNT {
+          emit RefilledFailed(address: address, reason: "Insufficient balance to refuel")
+          return
+        }
+      } else {
+        emit RefilledFailed(address: address, reason: "Couldn't borrow flowToken balance")
+        return
+      }
+
+      // 0.06 = 6MB of storage, or ~20k NBA TS moments
+      privateForwardingSenderRef!.sendPrivateTokens(address,tokens:<-vaultRef!.withdraw(amount: REFUEL_AMOUNT))
+      self.addRefilledAccount(address)
+      emit Refuelled(address)
+    } else {
+      emit RefilledFailed(address: address, reason: "Address is not below StorageRentRefillThreshold")
+    }
+  }
+
+  /// checkEligibility
+  ///
+  /// @param address: Address to check eligibility on
+  /// @return Boolean valued based on if the provided address is below the storage threshold
+  pub fun checkEligibility(_ address: Address): Bool {
+    if self.RefilledAccountInfos[address] != nil && getCurrentBlock().height - self.RefilledAccountInfos[address]!.atBlock < self.RefillRequiredBlocks {
+       return false
+    }
+    let acct = getAccount(address)
+    var high: UInt64 = acct.storageCapacity
+    var low: UInt64 = acct.storageUsed
+    if high < low {
+      high <-> low
+    }
+
+    if high - low >= self.StorageRentRefillThreshold {
+      return false
+    }
+
+    return true
+  }
+
+  /// addRefilledAccount
+  ///
+  /// @param address: Address to add to RefilledAccounts/RefilledAccountInfos
+  access(contract) fun addRefilledAccount(_ address: Address) {
+    if self.RefilledAccountInfos[address] != nil {
+      self.RefilledAccounts.remove(at: self.RefilledAccountInfos[address]!.index)
+    }
+
+    self.RefilledAccounts.append(address)
+    self.RefilledAccountInfos[address] = RefilledAccountInfo(self.RefilledAccounts.length-1, getCurrentBlock().height)
+  }
+
+  /// cleanExpiredRefilledAccounts
+  /// public method to clean up expired accounts based on current block height
+  ///
+  /// @param batchSize: Int to set the batch size of the cleanup
+  pub fun cleanExpiredRefilledAccounts(_ batchSize: Int) {
+    var index = 0
+    while index < batchSize && self.RefilledAccounts.length > index {
```

Comment:
This is incorrect as it would only allow running the loop half times than the given batchSize.

If the given batchSize is 10, this loop will run 5 times because of `self.RefilledAccounts.length` also decreases when you remove the index.

Reply:
nice catch!! working on a fix

Reply:
@satyamakgec updated method to the below, thoughts?

```
  pub fun cleanExpiredRefilledAccounts(_ batchSize: Int) {
    var index = 0
    var refilledAccountsToCleanup: [Address] = [];
    var refilledAccountsLength = self.refilledAccounts.length
    while index < batchSize && index < refilledAccountsLength {
      if self.refilledAccountInfos[self.refilledAccounts[index]] != nil &&
        getCurrentBlock().height - self.refilledAccountInfos[self.refilledAccounts[index]]!.atBlock > self.refillRequiredBlocks {
        refilledAccountsToCleanup.append(self.refilledAccounts[index])
      }
      index = index + 1
    }

    for account in refilledAccountsToCleanup {
        if let idx = self.refilledAccounts.firstIndex(of: account) {
          self.refilledAccounts.remove(at: idx)
          self.refilledAccountInfos.remove(key: account)
        }
    }
  }
```

Reply:
We can avoid the below `for` loop as we can move the removed statements of the array in the while loop. We need only these 2 statements. 

```
var refilledAccountsLength = self.refilledAccounts.length
while index < batchSize && index < refilledAccountsLength {
 ```